HPCC-25219 add placement configuration support.

xwang2713 opened the pull request:
This PR replaces https://github.com/hpcc-systems/HPCC-Platform/pull/14582
The configuration includes followings per pod,target or type:
nodeSelector,tolerations,affinity and schedulerName.

\#\# placements
\#\# The placement is responsible for finding the best Node for a Pod. It can be configured
\#\# through placement to include an array of objects to configure the Kubernetes Scheduler.
\#\# It must have a "pods" list which tells it which pod the settings will be applied to.
\#\# The synatx is:
\#\#   placements:
\#\#   - pods [list]
\#\#     placement:
\#\#       <supported configurations>
\#\#
\#\# The list item in "pods" can be one of the following:
\#\# 1) HPCC Systems component types in format: "type:<type name>". <type name> includes
\#\#    dali, esp, eclagent, eclccserver, roxie, thor. For example "type:esp"
\#\# 2) Target, the name of array item of above type in format "target:<target name>"
\#\#    For example "target:roxie", "target:thor".
\#\# 3) Pod, "Deployment" metadata name, which usually should be from the name of the array
\#\#    item of a type. For example, "eclwatch", "mydali", "thor-thoragent"..
\#\# 4) Job name regular express:  For example "compile-" or "compile-.*" or exact match "^compile-.*$"
\#\#
\#\# Supported configurations under each "placement"
\#\# 1) nodeSelector
\#\# 2) taints/tolerations
\#\# 3) affinity
\#\#    There is no schema check for the content of affinity. Reference
\#\#    https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/
\#\# 4) schedulerName: profile names. "affinity" defined in scheduler profile  requires
\#\#    Kubernetes 1.20.0 beta and later releases
\#\#
\#\# Examples:
\#placements:
\#- pods: ["eclwatch","roxie-workunit","^compile-.*$","mydali"]
\#  placement:
\#    nodeSelector:
\#      name: npone
\#- pods: ["thorworker-", "thor-thoragent", "thormanager-","thor-eclagent","hthor-"]
\#  placement:
\#    nodeSelector:
\#      name: nptwo
\#    tolerations:
\#    - key: "name"
\#      operator: "Equal"
\#      value: "nptwo"
\#      effect: "NoSchedule"
\#- pods: ["thorworker-.*"]
\#  placement:
\#    affinity:
\#      nodeAffinity:
\#        requiredDuringSchedulingIgnoredDuringExecution:
\#          nodeSelectorTerms:
\#          - matchExpressions:
\#            - key: kubernetes.io/e2e-az-name
\#              operator: In
\#              values:
\#              - e2e-az1
\#              - e2e-az2
\#- pods: ["target:roxie"]
\#  placement:
\#    schedulerName: "my-scheduler"
\#\# The settings will be applied to all thor pods/jobs and myeclccserver pod and job
\#- pods: ["target:myeclccserver", "type:thor"]
\#  placement:
\#    nodeSelector:
\#      app: "tf-gpu"
\#    tolerations:
\#    - key: "gpu"
\#      operator: "Equal"
\#      value: "true"
\#      effect: "NoSchedule"

Signed-off-by: Xiaoming Wang xiaoming.wang@lexisnexis.com


<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->
1) uncomment "placements" examples in values.yaml.
   Use "helm template" to generate yaml content to a file.
   Check the contents
2) Only uncomment "nodeSelector" and "tolerations" in "placement" of the values.yaml.
    Start AKS by running start.sh in https://github.com/gfortil/HPCC-Tests/tree/main/Azure/Kubernetes/Scheduler-Test/aks-np-1
    Deploy HPCC cluster 
    Validate the node assigning according the nodeSelector and tolerations settings
    Delete AKS cluster by running delete-resource-group.sh in https://github.com/gfortil/HPCC-Tests/tree/main/Azure/Kubernetes/Scheduler-Test/aks-np-1

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Process of PR-14665, label: HPCC-25219-placements is skipped.
The reason is: It is related to containerised environment only.
Commit ID: f212c62a80f5961ecb2bb7882d9e2d7267e59b1a
OS: centos 7.7.1908 (Linux 3.10.0-327.10.1.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: hpcc-platform-dev-el7-smoketest-2

richardkchapman commented on helm/hpcc/values.yaml:
```diff
@@ -49,6 +49,75 @@ security:
+## The placement is responsible for finding the best Node for a Pod. It can be configured
+## through placement to include an array of objects to configure the Kubernetes Scheduler.
+## It must have a "pods" list which tells it which pod the settings will be applied to.
+## The synatx is:
```
Typo

xwang2713 replied:
Fixed

richardkchapman commented on helm/hpcc/values.yaml:
```diff
@@ -49,6 +49,75 @@ security:
     extern: "allow"
     datafile: "allow"
 
+## placements
```
As a general comment (not specific to this PR but we need to think about it), the values.yaml is starting to get very large. Do we want so much in-line documentation? Is this the best location for the information, and is it starting to make the values.yaml hard to read/follow? Is there a better solution?

xwang2713 replied:
Another solution is to instruct user:
1) helm template to generate yaml file
2) use kustomize or similar tool to patch whatever changes 
3) kubectl apply/create to deploy the cluster
4) kubectl patch to change dynamic settings

richardkchapman replied:
I don't object to the values file being the place where we set this information.

I'm questioning whether the default values file is the best place to document this information

xwang2713 replied:
Maybe just point a document for user and move all instructions including examples to the documentation?. I am OK with it.

jakesmith replied:
I think that's sensible, but there are pre-existing sections that could also benefit from moving out into a document.
Let's leave this as it is for now, and tackle tidying up/referencing doc(s) from values.yaml in a separate JIRA/PR

xwang2713 replied:
Or put placements instructions to a file, such as Placements.md

jakesmith commented:
@xwang2713 - something may have merged since you pushed, but is now conflicting. Can you rebase ?

HPCCSmoketest commented:
Process of PR-14665, label: HPCC-25219-placements is skipped.
The reason is: It is related to containerised environment only.
Commit ID: ebdbfe0aabe2cd1a684bd27523c70cc7074c9364
OS: centos 7.7.1908 (Linux 3.10.0-327.10.1.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: hpcc-platform-dev-el7-smoketest-2

xwang2713 commented:
Fixed typo and rebase with master

jakesmith commented:
@xwang2713 - still appears to be basing problem (see 9 commits).

xwang2713 commented:
Rebase again as @jakesmith suggested

HPCCSmoketest commented:
Process of PR-14665, label: HPCC-25219-placements is skipped.
The reason is: It is related to containerised environment only.
Commit ID: d380bb407cd3fa713d4e71eab5d5fcf6c190bd9d
OS: centos 7.7.1908 (Linux 3.10.0-327.10.1.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: hpcc-platform-dev-el7-smoketest-2

xwang2713 commented:
Wait. Still have rebase problem

xwang2713 committed:
HPCC-25219 add placement configuration support
The configuration includes followings per pod,target or type:
nodeSelector, tolerations, affinity and schedulerName

HPCCSmoketest commented:
Process of PR-14665, label: HPCC-25219-placements is skipped.
The reason is: It is related to containerised environment only.
Commit ID: dd073ee7b8dbfc5c6bb6d0157207ed700c10b7c0
OS: centos 7.7.1908 (Linux 3.10.0-327.10.1.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: hpcc-platform-dev-el7-smoketest-2

xwang2713 commented:
Rebased.
Move instruction and examples to file docs/placements.md

jakesmith commented on helm/hpcc/templates/eclccserver.yaml:
```diff
@@ -3,6 +3,7 @@ EclccServer configmap
 Pass in dict with root and me
 */}}
 {{- define "hpcc.eclccServerConfigMap" -}}
+{{- $compileJobName := printf "compile-%%jobname" }}
```
This change doesn't look necessary, the printf isn't acutally doing anything here (no arguments).
Could revert this and line 27 to way it was.

jakesmith reviewed (approved):
@xwang2713 - looks good to me.
